src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java

```diff
@@ -45,20 +43,23 @@ public OtpUserController(String apiPrefix) {
 
     @Override
     OtpUser preCreateHook(OtpUser user, Request req) {
-        // Check API key and assign user to appropriate third-party application. Note: this is only relevant for
-        // instances of otp-middleware running behind API Gateway.
-        String apiKey = req.headers("x-api-key");
-        ApiUser apiUser = Persistence.apiUsers.getOneFiltered(Filters.eq("apiKeys.value", apiKey));
-        if (apiUser != null) {
-            // If API user found, assign to new OTP user.
-            user.applicationId = apiUser.id;
-        } else {
-            // If API user not found, report to Bugsnag for further investigation.
-            BugsnagReporter.reportErrorToBugsnag(
-                "OTP user created with API key that is not linked to any API user",
-                apiKey,
-                new IllegalArgumentException("API key not linked to API user.")
-            );
+        RequestingUser requestingUser = Auth0Connection.getUserFromRequest(req);
+        // If an Api user is present it is assumed an attempt is being made to create an OTP user.
+        if (requestingUser.apiUser != null) {
+            String apiKeyFromHeader = req.headers("x-api-key");
+            // Check API key and assign user to appropriate third-party application. Note: this is only relevant for
+            // instances of otp-middleware running behind API Gateway.
+            if (apiKeyFromHeader != null && requestingUser.apiUser.hasToken(apiKeyFromHeader)) {
+                // If third party and using their own api key, assign to new OTP user.
+                user.applicationId = requestingUser.apiUser.id;
+            } else {
+                // If API user not found, log message and halt.
+                logMessageAndHalt(
+                    req,
+                    HttpStatus.FORBIDDEN_403,
+                    "Attempting to create OTP user with API key that is not linked to calling third party",
+                    new IllegalArgumentException("API key not linked to API user."));
+            }
```

Comment:
This whole check doesn't really make sense to me here... shouldn't it be at a lower level (i.e., in Auth0Connection)?

Reply:
Moved the validation part into `Auth0Connection#linkApiKeyToApiUser`.